flatdata-rs/lib/src/tarstorage.rs

```diff
@@ -0,0 +1,128 @@
+use crate::storage::{ResourceStorage, StorageHandle, Stream};
+
+use memmap2::Mmap;
+use std::{
+    collections::HashMap,
+    fs::File,
+    io,
+    ops::Range,
+    path::{Path, PathBuf},
+    slice,
+    sync::Arc,
+};
+
+/// Internal storage of file entries in tar archive.
+#[derive(Debug)]
+struct MemoryMappedTarArchiveStorage {
+    archive_map: Mmap,
+    file_ranges: HashMap<PathBuf, Range<usize>>,
+}
+
+impl MemoryMappedTarArchiveStorage {
+    pub fn new(tar_path: &Path) -> Result<Self, io::Error> {
+        let file = File::open(tar_path)?;
+        let mut archive = tar::Archive::new(file);
+
+        let file_ranges = archive
+            .entries()?
+            .map(|entry| {
+                let entry = entry?;
+                let path = entry.path()?;
+                let path = if let Ok(stripped_path) = path.strip_prefix(".") {
+                    stripped_path.to_path_buf()
+                } else {
+                    path.to_path_buf()
+                };
+                let offset = entry.raw_file_position() as usize;
```

Comment:
Is there a check anywhere that the file is actually stored in a consecutive way?

Reply:
There are AFAIK two cases in which a file is not stored contiguously:
1. Sparse files (files with holes in them)
2. Multivolume files

I added a check to fail in case we encounter a sparse file.
Multivolume files are currently not supported by tar-rs.